### PR TITLE
Enable rtx sources for streams

### PR DESCRIPTION
**Pull request by bbaldino**

### CODE REVIEW

File: modules/xmpp/RtxModifier.js [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+ * Begin helper functions
+ */
+/**
+ * Given a videoMLine, returns a list of the video
+ *  ssrcs (those used to actually send video, not
+ *  any associated secondary streams)
+ * @param videoMLine media line object from transform.parse
+ * @returns list of ssrcs (as strings)
+ */
+function getPrimaryVideoSsrcs (videoMLine) {
```
Comment (paweldomas) [inline modules/xmpp/RtxModifier.js]: It feels like those helper functions could be extracted to some utility class, so that they would be more visible for someone unfamiliar with RtcModifier. The same applies to the SdpConsistency. Something like SdpTransformUtil that would gather helpers for operating with the 'sdp-transform' lib.

File: modules/xmpp/RtxModifier.js [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+        .filter((ssrc, index, array) => array.indexOf(ssrc) === index);
+
+    if (videoMLine.ssrcGroups) {
+        videoMLine.ssrcGroups.forEach((ssrcGroupInfo) => {
+            // Right now, FID groups are the only ones we parse to 
+            //  disqualify streams.  If/when others arise we'll
+            //  need to add support for them here
+            if (ssrcGroupInfo.semantics === "FID") {
+                // secondary FID streams should be filtered out
+                let secondarySsrc = ssrcGroupInfo.ssrcs.split(" ")[1];
```
Comment (paweldomas) [inline modules/xmpp/RtxModifier.js]: I see that split(" ") then parsing to number is duplicated in many places. Can we wrap that into some util/class ?

File: modules/xmpp/RtxModifier.js [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+            if (group.semantics === "FID") {
+                let primarySsrc = group.ssrcs.split(" ")[0];
+                let rtxSsrc = group.ssrcs.split(" ")[1];
+                if (!ssrcExists(primarySsrc)) {
+                    ssrcsToRemove.push(rtxSsrc);
+                }
+            }
+        });
+        videoMLine.ssrcs = videoMLine.ssrcs
+            .filter(ssrc => ssrcsToRemove.indexOf(ssrc.id + "") === -1);
```
Comment (paweldomas) [inline modules/xmpp/RtxModifier.js]: I see that there is this problem of converting between strings and numbers when dealing with sdp-transform. Maybe we should review our strategy to store SSRCs as strings instead ?

Comment (bbaldino) [inline modules/xmpp/RtxModifier.js]: yeah this does get annoying.  for SdpConsistency i changed everything that returned them or took them in as args to use numbers and always convert away from string.  here there's a bit of a mix because always forcing them felt awkward.  i think part of it comes from the fact that the 'id' field in the sdp ssrc line uses a number, but the ssrc group field has the line as a string of space-separated ssrc strings

File: modules/xmpp/RtxModifier.js (hunk 2 above)
Comment (bbaldino) [inline modules/xmpp/RtxModifier.js]: yeah i think it'd be a good idea to write a 'parse ssrcs from group' helper method or something.  i'll do that.

Comment (bbaldino) [inline modules/xmpp/RtxModifier.js]: fixed this

**Comment by paweldomas** [comment]
I find SDPUtil slightly confusing, because they are not related with "sdp-transform" lib and mix operation on objects coming from it with manual SDP parsing, but it's fine as long as the arguments are documented, so I'm going to merge it anyway.

File: modules/xmpp/RtxModifier.js [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+function getRtxSsrc (videoMLine, primarySsrc) {
+    if (videoMLine.ssrcGroups) {
+        let fidGroup = videoMLine.ssrcGroups.find(group => {
+            if (group.semantics === "FID") {
+                let groupPrimarySsrc = SDPUtil.parseGroupSsrcs(group)[0];
+                return groupPrimarySsrc === primarySsrc;
+            }
+        });
+        if (fidGroup) {
+          return parseInt(fidGroup.ssrcs.split(" ")[1]);
```
Comment (paweldomas) [inline modules/xmpp/RtxModifier.js]: Would it be suitable to call SDPUtil.parseGroupSsrcs here as well ? (sorry for being so picky :P)

File: modules/xmpp/RtxModifier.js [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+export default class RtxModifier {
+    /**
+     * Constructor
+     */
+    constructor () {
+        /**
+         * Map of video ssrc to corresponding RTX
+         *  ssrc
+         */
+        this.correspondingRtxSsrcs = {};
```
Comment (saghul) [inline modules/xmpp/RtxModifier.js]: nitpick: maybe use a Map? (ES6 ALL THE THINGS)

File: modules/xmpp/RtxModifier.spec.js [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+
+        it ("should re-use the same rtx ssrc for a primary ssrc it's seen before", function() {
+          // Have rtxModifier generate an rtx ssrc via modifyRtxSsrcs.  Then call it again
+          //  with the same primary ssrc in the sdp (but no rtx ssrc).  It should use
+          //  the same rtx ssrc as before.
+          let newSdpStr = this.rtxModifier.modifyRtxSsrcs(this.transform.write(this.singleVideoSdp));
+          let newSdp = transform.parse(newSdpStr);
+
+          let fidGroup = getVideoGroups(newSdp, "FID")[0];
+          let fidGroupRtxSsrc = parseInt(fidGroup.ssrcs.split(" ")[1]);
```
Comment (paweldomas) [inline modules/xmpp/RtxModifier.spec.js]: "SDPUtil parse groups" should be used here as well ? (please fix also other occurrences in this file)

File: modules/xmpp/TraceablePeerConnection.js [hunk 7]
```diff
@@ -583,18 +616,35 @@ TraceablePeerConnection.prototype.getStats = function(callback, errback) {
...
             group: {ssrcs: ssrcInfo.ssrcs.join(" "), semantics: "SIM"}});
-        return ssrcInfo;
+        ssrcInfo;
     } else {
-        return {ssrcs: [RandomUtil.randomInt(1, 0xffffffff)], groups: []};
+        ssrcInfo = {ssrcs: [SDPUtil.generateSsrc()], groups: []};
+    }
+    if (!this.session.room.options.disableRtx) {
+        let currNumSsrcs = ssrcInfo.ssrcs.length;
+        for (let i = 0; i < currNumSsrcs; ++i) {
```
Comment (saghul) [inline modules/xmpp/TraceablePeerConnection.js]: for..of

File: modules/xmpp/RtxModifier.js [hunk 8]
```diff
@@ -0,0 +1,276 @@
+import { getLogger } from "jitsi-meet-logger";
+const logger = getLogger(__filename);
+import * as transform from 'sdp-transform';
+var SDPUtil = require("./SDPUtil");
```
Comment (saghul) [inline modules/xmpp/RtxModifier.js]: const?

**Comment by saghul** [comment]
@bbaldino left a couple of drive-by comments.

File: modules/xmpp/RtxModifier.spec.js [hunk 9]
```diff
@@ -0,0 +1,263 @@
+/*eslint-disable max-len*/
+/*jshint maxlen:false*/
+var RtxModifier = require("./RtxModifier.js");
```
Comment (paweldomas) [inline modules/xmpp/RtxModifier.spec.js]: Can we stick with ES6 "import" with the new code ? (please fix all occurrences)
In general we should try to use ES6 with the new code. I'm not sure if that works with the tests, does it ?

File: modules/xmpp/RtxModifier.spec.js [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+};
+
+var getVideoGroups = function (parsedSdp, groupSemantics) {
+  let videoMLine = parsedSdp.media.find(m => m.type === "video");
+  videoMLine.ssrcGroups = videoMLine.ssrcGroups || [];
+  return videoMLine.ssrcGroups
+    .filter(g => g.semantics === groupSemantics);
+};
+
+describe ("RtxModifier", function() {
```
Comment (paweldomas) [inline modules/xmpp/RtxModifier.spec.js]: can we replace all function() with () => if possible (not sure if that works with the tests)

File: modules/xmpp/SDPUtil.js [hunk 11]
```diff
@@ -1,7 +1,7 @@
 import {getLogger} from "jitsi-meet-logger";
 const logger = getLogger(__filename);
 var RTCBrowserType = require("../RTC/RTCBrowserType");
-
+var RandomUtil = require('../util/RandomUtil');
```
Comment (paweldomas) [inline modules/xmpp/SDPUtil.js]: please use "import" instead

File: modules/xmpp/RtxModifier.js (hunk 4 above)
Comment (bbaldino) [inline modules/xmpp/RtxModifier.js]: guess i missed this one...fixed.

File: modules/xmpp/RtxModifier.spec.js (hunk 6 above)
Comment (bbaldino) [inline modules/xmpp/RtxModifier.spec.js]: fixed

File: modules/xmpp/TraceablePeerConnection.js (hunk 7 above)
Comment (bbaldino) [inline modules/xmpp/TraceablePeerConnection.js]: fixed

Comment (bbaldino) [inline modules/xmpp/TraceablePeerConnection.js]: i find the syntax of for..of very unfortunate btw:
`for (let primarySsrc of ssrcInfo.ssrcs) {`
doesn't read very nicely to me.

File: modules/xmpp/RtxModifier.js (hunk 8 above)
Comment (bbaldino) [inline modules/xmpp/RtxModifier.js]: fixed

File: modules/xmpp/RtxModifier.spec.js (hunk 9 above)
Comment (bbaldino) [inline modules/xmpp/RtxModifier.spec.js]: fixed

File: modules/xmpp/TraceablePeerConnection.js (hunk 7 above)
Comment (paweldomas) [inline modules/xmpp/TraceablePeerConnection.js]: in this particular case it does read better when a ssrc, becomes a primary SSRC, but I don't have strong opinion

File: modules/xmpp/RtxModifier.spec.js (hunk 10 above)
Comment (bbaldino) [inline modules/xmpp/RtxModifier.spec.js]: unfortunately (or fortunately, maybe?) jasmine passes some context in a passed 'this', so some of the functions (`beforeEach` and `it`) are best using the old-style functions.  so our option is to use arrow syntax in some places but not others, or just be consistent with old-style functions everywhere.  what do you think?

File: modules/xmpp/SDPUtil.js (hunk 11 above)
Comment (bbaldino) [inline modules/xmpp/SDPUtil.js]: fixed

File: modules/xmpp/RtxModifier.spec.js (hunk 10 above)
Comment (paweldomas) [inline modules/xmpp/RtxModifier.spec.js]: If that's causing issues leave it as is.

File: modules/xmpp/RtxModifier.js (hunk 5 above)
Comment (bbaldino) [inline modules/xmpp/RtxModifier.js]: fixed.  also not crazy about the `set` syntax of `Map`.  to me
`myMap[key] = value;`
is clearer than
`myMap.set(key, value);`
but oh well.

File: modules/xmpp/RtxModifier.spec.js (hunk 10 above)
Comment (bbaldino) [inline modules/xmpp/RtxModifier.spec.js]: i agree...i think being consistent (always use old-style functions in tests) will be more straightforward and less error prone.

**Comment by bbaldino** [comment]
@paweldomas let me know if this looks good, but please don't merge it quite yet.

File: modules/xmpp/TraceablePeerConnection.js (hunk 7 above)
Comment (bbaldino) [inline modules/xmpp/TraceablePeerConnection.js]: actually i had to change it back...we add to the list being iterated over here, so i explicitly grabbed the original length so we only iterated over the original items.  i added a comment here too to make that more clear.

**Comment by bbaldino** [comment]
ps @paweldomas i agree with your sentiment about SDPUtil.  it's a bit of a catch-all at the moment, and honestly i think there is a lot of sdp functionality that would be useful as a lib of its own.  maybe that would build on top of sdp-transform or not, i dunno.

**Comment by paweldomas** [comment]
> ps @paweldomas i agree with your sentiment about SDPUtil. it's a bit of a catch-all at the moment, and honestly i think there is a lot of sdp functionality that would be useful as a lib of its own. maybe that would build on top of sdp-transform or not, i dunno.

Yes, I agree. Have been thinking about the same (possibly using sdp-transform on top). Like I wrote it is fine as long as it's documented (the methods are still SDP utilities).

**Comment by paweldomas** [comment]
Looks good, I don't want to hold it off anymore. Is it ready to be merged now ?

File: modules/xmpp/RtxModifier.js [hunk 12]
```diff
@@ -0,0 +1,276 @@
+import { getLogger } from "jitsi-meet-logger";
+const logger = getLogger(__filename);
+import * as transform from 'sdp-transform';
+const SDPUtil = require("./SDPUtil");
```
Comment (hristoterezov) [inline modules/xmpp/RtxModifier.js]: Why not `import * as SDPUtil from "./SDPUtil"`?

File: modules/xmpp/RtxModifier.spec.js [hunk 13]
```diff
@@ -0,0 +1,268 @@
+/*eslint-disable max-len*/
```
Comment (hristoterezov) [inline modules/xmpp/RtxModifier.spec.js]: Why do we have mixed requires with imports and vars with const/let in this file?

File: modules/xmpp/SampleSdpStrings.js [hunk 14]
```diff
@@ -0,0 +1,162 @@
+/*eslint-disable max-len*/
```
Comment (hristoterezov) [inline modules/xmpp/SampleSdpStrings.js]: Can't we use let/const in this file?

File: modules/xmpp/SampleSdpStrings.js [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+"a=ssrc-group:FID 1089111804 2963867077\r\n" +
+"a=ssrc-group:SIM 1757014965 1479742055 1089111804\r\n" +
+"a=rtcp-mux\r\n";
+
+var simulcastSdp = baseSessionSdp + baseAudioMLineSdp + simulcastVideoMLineSdp + baseDataMLineSdp;
+var simulcastRtxSdp = baseSessionSdp + baseAudioMLineSdp + simulcastRtxVideoMLineSdp + baseDataMLineSdp;
+var plainVideoSdp = baseSessionSdp + baseAudioMLineSdp + plainVideoMLineSdp + baseDataMLineSdp;
+var rtxVideoSdp = baseSessionSdp + baseAudioMLineSdp + rtxVideoMLineSdp + baseDataMLineSdp;
+
+exports.simulcastSdp = transform.parse(simulcastSdp);
```
Comment (hristoterezov) [inline modules/xmpp/SampleSdpStrings.js]: Why not export `export const simulcastSdp = transform.parse(simulcastSdp);`?

PS: The question is valid for the next lines also.

File: modules/xmpp/SdpConsistency.js [hunk 16]
```diff
@@ -1,4 +1,5 @@
 import * as transform from 'sdp-transform';
+var SDPUtil = require("./SDPUtil");
```
Comment (hristoterezov) [inline modules/xmpp/SdpConsistency.js]: Why not import * as SDPUtil from "./SDPUtil"?

File: modules/xmpp/SampleSdpStrings.js [hunk 17]
```diff
@@ -0,0 +1,162 @@
+/*eslint-disable max-len*/
+/*jshint maxlen:false*/
+var transform = require('sdp-transform');
```
Comment (hristoterezov) [inline modules/xmpp/SampleSdpStrings.js]: Why not using import?

File: modules/xmpp/SampleSdpStrings.js [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+
+var baseSessionSdp = "" +
```
Comment (hristoterezov) [inline modules/xmpp/SampleSdpStrings.js]: Could you please add comments for every variable?

File: modules/xmpp/SdpConsistency.js [hunk 19]
```diff
@@ -141,36 +118,28 @@ export default class SdpConsistency {
...
             }
-            let newPrimaryRtxSsrc =
-                getRtxSsrc(videoMLine, newPrimarySsrc);
             if (!this.cachedPrimarySsrc) {
                 this.cachedPrimarySsrc = newPrimarySsrc;
-                this.cachedPrimaryRtxSsrc = newPrimaryRtxSsrc;
-                console.log("Sdp-consistency caching primary ssrc " +
-                    this.cachedPrimarySsrc + " and rtx " +
-                    this.cachedPrimaryRtxSsrc);
+                console.log("Sdp-consistency caching primary ssrc " +
```
Comment (hristoterezov) [inline modules/xmpp/SdpConsistency.js]: I see that this is old code but I was wondering why we are using console.log instead of logger.log? Do you have any idea?

If there isn't reason for this maybe it's good moment to change them to logger since you are changing them. WDYT?

File: modules/xmpp/RtxModifier.js [hunk 20]
```diff
@@ -0,0 +1,276 @@
...
+        logger.info("Setting ssrc cache to ", ssrcMapping);
+        this.correspondingRtxSsrcs = ssrcMapping;
+    }
+
+    /**
+     * Adds RTX ssrcs for any video ssrcs that don't
+     *  already have them.  If the video ssrc has been
+     *  seen before, and already had an RTX ssrc generated,
+     *  the same RTX ssrc will be used again.
+     * @param {string} sdpStr sdp in raw string format
```
Comment (hristoterezov) [inline modules/xmpp/RtxModifier.js]: I believe there should be `@returns`.

File: modules/xmpp/RtxModifier.js [hunk 21]
```diff
@@ -0,0 +1,276 @@
...
+        });
+        return transform.write(parsedSdp);
+    }
+
+    /**
+     * Remove all reference to any rtx ssrcs that 
+     *  don't correspond to the primary stream.
+     * Must be called *after* any simulcast streams
+     *  have been imploded
+     * @param {string} sdpStr sdp in raw string format
```
Comment (hristoterezov) [inline modules/xmpp/RtxModifier.js]: I believe there should be `@returns`.

File: modules/xmpp/RtxModifier.spec.js [hunk 22]
```diff
@@ -0,0 +1,268 @@
...
+/*jshint maxlen:false*/
+import RtxModifier from "./RtxModifier.js";
+var SampleSdpStrings = require("./SampleSdpStrings.js");
+var transform = require('sdp-transform');
+var SDPUtil = require("./SDPUtil.js");
+
+var numVideoSsrcs = (parsedSdp) => {
```
Comment (hristoterezov) [inline modules/xmpp/RtxModifier.spec.js]: 1) Missing JSDoc
2) Why not `function numVideoSsrcs()...`?

File: modules/xmpp/RtxModifier.spec.js [hunk 23]
```diff
@@ -0,0 +1,268 @@
...
+  let videoMLine = parsedSdp.media.find(m => m.type === "video");
+  return videoMLine.ssrcs
+    .map(ssrcInfo => ssrcInfo.id)
+    .filter((ssrc, index, array) => array.indexOf(ssrc) === index)
+    .length;
+};
+
+var getPrimaryVideoSsrc = function (parsedSdp) {
```
Comment (hristoterezov) [inline modules/xmpp/RtxModifier.spec.js]: 1) Missing JSDoc
2) Why not `function getPrimaryVideoSsrc()...`?

File: modules/xmpp/RtxModifier.spec.js [hunk 24]
```diff
@@ -0,0 +1,268 @@
...
+  let videoMLine = parsedSdp.media.find(m => m.type === "video");
+  return parseInt(SDPUtil.parsePrimaryVideoSsrc(videoMLine));
+};
+
+// Only handles parsing 2 scenarios right now:
+// 1) Single video ssrc
+// 2) Multiple video ssrcs in a single simulcast group
+var getPrimaryVideoSsrcs = function (parsedSdp) {
```
Comment (hristoterezov) [inline modules/xmpp/RtxModifier.spec.js]: 1) Missing JSDoc
2) Why not `function getPrimaryVideoSsrcs()...`?

File: modules/xmpp/RtxModifier.spec.js [hunk 25]
```diff
@@ -0,0 +1,268 @@
...
+    let simGroups = getVideoGroups(parsedSdp, "SIM");
+    if (simGroups.length > 1) {
+      return;
+    }
+    let simGroup = simGroups[0];
+    return SDPUtil.parseGroupSsrcs(simGroup);
+  }
+};
+
+var getVideoGroups = function (parsedSdp, groupSemantics) {
```
Comment (hristoterezov) [inline modules/xmpp/RtxModifier.spec.js]: 1) Missing JSDoc
2) Why not `function getVideoGroups()...`?

File: modules/xmpp/SDPUtil.js [hunk 26]
```diff
@@ -362,7 +362,69 @@ var SDPUtil = {
         line += ' ';
         line += cand.getAttribute('generation') || '0';
         return line + '\r\n';
-    }
+    },
+
+    parsePrimaryVideoSsrc: function(videoMLine) {
```
Comment (hristoterezov) [inline modules/xmpp/SDPUtil.js]: Missing JSDoc.

File: modules/xmpp/SDPUtil.js [hunk 27]
```diff
@@ -362,7 +362,69 @@ var SDPUtil = {
...
+                    .find(group => group.semantics === "SIM");
+                if (simGroup) {
+                    primarySsrc = simGroup.ssrcs.split(" ")[0];
+                }
+            }
+        }
+        return primarySsrc;
+    },
+
+    generateSsrc: function() {
```
Comment (hristoterezov) [inline modules/xmpp/SDPUtil.js]: Missing JSDoc.

File: modules/xmpp/SDPUtil.js [hunk 28]
```diff
@@ -362,7 +362,69 @@ var SDPUtil = {
...
+        return RandomUtil.randomInt(1, 0xffffffff);
+    },
+
+    getSsrcAttribute: function (mLine, ssrc, attributeName) {
```
Comment (hristoterezov) [inline modules/xmpp/SDPUtil.js]: Missing JSDoc.

**Comment by hristoterezov** [comment]
There are a lot of places where you have used `let` but actually the variable is never changed, so could you please change those variables to `const`?

File: modules/xmpp/RtxModifier.js [hunk 29]
```diff
@@ -0,0 +1,276 @@
...
+ * Begin helper functions
+ */
+/**
+ * Given a videoMLine, returns a list of the video
+ *  ssrcs (those used to actually send video, not
+ *  any associated secondary streams)
+ * @param {object} videoMLine media line object from transform.parse
+ * @returns {list<string>} list of primary video ssrcs
+ */
+function getPrimaryVideoSsrcs (videoMLine) {
```
Comment (hristoterezov) [inline modules/xmpp/RtxModifier.js]: If we have simulcast enabled isn't it going to return also all simulcast ssrcs?

File: modules/xmpp/RtxModifier.js [hunk 30]
```diff
@@ -0,0 +1,276 @@
...
+ *  corresponding rtx ssrc
+ * @returns {number} the rtx ssrc (or undefined if there isn't one)
+ */
+function getRtxSsrc (videoMLine, primarySsrc) {
+    if (videoMLine.ssrcGroups) {
+        let fidGroup = videoMLine.ssrcGroups.find(group => {
+            if (group.semantics === "FID") {
+                let groupPrimarySsrc = SDPUtil.parseGroupSsrcs(group)[0];
+                return groupPrimarySsrc === primarySsrc;
+            }
```
Comment (hristoterezov) [inline modules/xmpp/RtxModifier.js]: Maybe it's good to add `return false;` by default. WDYT?

Comment (hristoterezov) [inline modules/xmpp/RtxModifier.js]: Also you can save one `SDPUtil.parseGroupSsrcs` call if you store the result from it.

File: modules/xmpp/TraceablePeerConnection.js [hunk 31]
```diff
@@ -317,7 +318,29 @@ TraceablePeerConnection.prototype.addStream = function (stream, ssrcInfo) {
         this.peerconnection.addStream(stream);
     if (ssrcInfo && ssrcInfo.type === "addMuted") {
         this.sdpConsistency.setPrimarySsrc(ssrcInfo.ssrc.ssrcs[0]);
-        this.simulcast.setSsrcCache(ssrcInfo.ssrc.ssrcs);
+        let simGroup =
```
Comment (hristoterezov) [inline modules/xmpp/TraceablePeerConnection.js]: Isn't it going to be better to first check for simulcast and rtx enabled/disabled before doing all those operations?

File: modules/xmpp/RtxModifier.js (hunk 12 above)
Comment (bbaldino) [inline modules/xmpp/RtxModifier.js]: fixed

File: modules/xmpp/RtxModifier.js (hunk 20 above)
Comment (bbaldino) [inline modules/xmpp/RtxModifier.js]: fixed

File: modules/xmpp/RtxModifier.js (hunk 21 above)
Comment (bbaldino) [inline modules/xmpp/RtxModifier.js]: fixed

File: modules/xmpp/RtxModifier.spec.js (hunk 13 above)
Comment (bbaldino) [inline modules/xmpp/RtxModifier.spec.js]: fixed

File: modules/xmpp/RtxModifier.spec.js (hunk 22 above)
Comment (bbaldino) [inline modules/xmpp/RtxModifier.spec.js]: fixed (both parts)

File: modules/xmpp/RtxModifier.spec.js (hunk 23 above)
Comment (bbaldino) [inline modules/xmpp/RtxModifier.spec.js]: fixed (both parts)

File: modules/xmpp/RtxModifier.spec.js (hunk 24 above)
Comment (bbaldino) [inline modules/xmpp/RtxModifier.spec.js]: fixed (both parts)

File: modules/xmpp/RtxModifier.spec.js (hunk 25 above)
Comment (bbaldino) [inline modules/xmpp/RtxModifier.spec.js]: fixed (both parts)

File: modules/xmpp/SampleSdpStrings.js (hunk 17 above)
Comment (bbaldino) [inline modules/xmpp/SampleSdpStrings.js]: fixed

File: modules/xmpp/SampleSdpStrings.js (hunk 18 above)
Comment (bbaldino) [inline modules/xmpp/SampleSdpStrings.js]: fixed

File: modules/xmpp/SdpConsistency.js (hunk 19 above)
Comment (bbaldino) [inline modules/xmpp/SdpConsistency.js]: yeah i dunno...fixed it.

File: modules/xmpp/RtxModifier.js [hunk 32]
```diff
@@ -0,0 +1,276 @@
...
+
+    /**
+     * Adds RTX ssrcs for any video ssrcs that don't
+     *  already have them.  If the video ssrc has been
+     *  seen before, and already had an RTX ssrc generated,
+     *  the same RTX ssrc will be used again.
+     * @param {string} sdpStr sdp in raw string format
+     */
+    modifyRtxSsrcs (sdpStr) {
+        let parsedSdp = transform.parse(sdpStr);
```
Comment (hristoterezov) [inline modules/xmpp/RtxModifier.js]: Maybe it's good to limit the number of transform.parse calls. Maybe we should have one transformed copy in higher level which we can pass here and whenever it is needed. 

I think the same is valid for transform.write. WDYT?

I'm not sure if the refactoring related to this is not too big to be included into this PR. Maybe we should do this in separate PR.

File: modules/xmpp/SDPUtil.js (hunk 26 above)
Comment (bbaldino) [inline modules/xmpp/SDPUtil.js]: fixed

File: modules/xmpp/SDPUtil.js (hunk 27 above)
Comment (bbaldino) [inline modules/xmpp/SDPUtil.js]: fixed

File: modules/xmpp/SDPUtil.js (hunk 28 above)
Comment (bbaldino) [inline modules/xmpp/SDPUtil.js]: fixed

File: modules/xmpp/RtxModifier.js (hunk 29 above)
Comment (bbaldino) [inline modules/xmpp/RtxModifier.js]: yeah that's ok, that's what we want (since we'll add rtx streams for each one)

File: modules/xmpp/RtxModifier.js (hunk 30 above)
Comment (bbaldino) [inline modules/xmpp/RtxModifier.js]: the result will be `undefined` by default already, won't it?  and that's what i wanted in the case of it not finding one anyway

File: modules/xmpp/TraceablePeerConnection.js (hunk 31 above)
Comment (bbaldino) [inline modules/xmpp/TraceablePeerConnection.js]: yup good idea

Comment (bbaldino) [inline modules/xmpp/TraceablePeerConnection.js]: oh, well actually this is basically doing that by checking for the presence of the sim and fid groups

File: modules/xmpp/RtxModifier.js (hunk 32 above)
Comment (bbaldino) [inline modules/xmpp/RtxModifier.js]: hm, but this file only calls it twice?  once for each of the flows (modifyRtxSsrcs on the local side, and implodeRemoteRtxSsrcs on the remote side)

**Comment by bbaldino** [comment]
> There are a lot of places where you have used let but actually the variable is never changed, so could you please change those variables to const?

i've been using let but not const as i haven't seen it much.  i'll make an effort to try and get RtxModifier.js cleaned up, but probably won't go further than that for now...will get in the habit of using it by default and only falling back to `let` when necessary (i assume that's what we want to do?)

File: modules/xmpp/SdpConsistency.js (hunk 16 above)
Comment (bbaldino) [inline modules/xmpp/SdpConsistency.js]: fixed

File: modules/xmpp/SampleSdpStrings.js (hunk 14 above)
Comment (bbaldino) [inline modules/xmpp/SampleSdpStrings.js]: fixed

File: modules/xmpp/SampleSdpStrings.js (hunk 15 above)
Comment (bbaldino) [inline modules/xmpp/SampleSdpStrings.js]: fixed

File: modules/xmpp/TraceablePeerConnection.js (hunk 31 above)
Comment (hristoterezov) [inline modules/xmpp/TraceablePeerConnection.js]: I meant checking `this.session.room.options.disableRtx`
